### PR TITLE
[Merged by Bors] - feat(Algebra/Module): Use coercion from SemilinearEquivClass to SemilinearEquiv

### PR DESCRIPTION
**Pull request by 101damnations**
---

Add the coercion from types with a `SemilinearEquivClass` instance to linear equivs, a la #10758. 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Equiv.lean [hunk 1]
```diff
@@ -122,6 +122,19 @@ instance (priority := 100) [RingHomInvPair σ σ'] [RingHomInvPair σ' σ]
...
+
+/-- Reinterpret an element of a type of semilinear equivalences as a semilinear equivalence. -/
+@[coe]
+def semilinearEquiv [RingHomInvPair σ σ'] [RingHomInvPair σ' σ]
+    [EquivLike F M M₂] [SemilinearEquivClass F σ M M₂] (f : F) : M ≃ₛₗ[σ] M₂ :=
+  { (f : M ≃+ M₂), (f : M →ₛₗ[σ] M₂) with }
+
+/-- Reinterpret an element of a type of semilinear equivalences as a semilinear equivalence. -/
+instance instCoeToSemilinearEquiv [RingHomInvPair σ σ'] [RingHomInvPair σ' σ]
+    [EquivLike F M M₂] [SemilinearEquivClass F σ M M₂] : CoeHead F (M ≃ₛₗ[σ] M₂) where
```
Comment (fpvandoorn) [inline Mathlib/Algebra/Module/Equiv.lean:135]:
```suggestion
    [EquivLike F M M₂] [SemilinearEquivClass F σ M M₂] : CoeTC F (M ≃ₛₗ[σ] M₂) where
```
For `Equiv` and `MulEquiv` this was done using `CoeTC`, so I imagine we want that everywhere?

bors d+

Comment (101damnations) [reply Mathlib/Algebra/Module/Equiv.lean:135]: I don't really understand this stuff but judging by [this comment](https://github.com/leanprover-community/mathlib4/pull/10368#discussion_r1488584069) it looks like `CoeHead` is preferred here

**Comment by 101damnations** [comment]
So there's already a coercion from linear equivs to linear maps given by the projection `LinearEquiv.toLinearMap`, and the following fails if I put it where it makes sense (e.g. after line 202)
```
example (f : M ≃ₛₗ[σ] M₂) :
    f.toLinearMap = SemilinearMapClass.semilinearMap f := by
  with_reducible rfl
``` 
I don't really understand `with_reducible`, but I guess maybe I should remove
```
instance : Coe (M ≃ₛₗ[σ] M₂) (M →ₛₗ[σ] M₂) :=
  ⟨toLinearMap⟩
```
on line 153, and deal with the fall out?

**Comment by fpvandoorn** [comment]
Thanks for the link. In that case, we should try to fix the existing coercions.

bors merge

**Pull request closed without merging**